examples/nlp/mlm_and_finetune_with_bert.py

```diff
@@ -0,0 +1,290 @@
+
+        # set 10% to a random token
+        inp_mask_2random = inp_mask_2mask & (np.random.rand(*X.shape) < 1 / 9)
+        X_mlm[inp_mask_2random] = np.random.randint(
+            3, mask_token_id, inp_mask_2random.sum()
+        )
+
+        # prepare sample_weights to pass to .fit() method
+        sample_weights = np.ones(labels.shape)
+        sample_weights[labels == -1] = 0
+
+        # y_labels would be same as X i.e input tokens
+        y_labels = np.copy(X)
+
+        return X_mlm, y_labels, sample_weights
+
+
+classifier_train_data = IMDBReviewsDataGenerator(
+    data["text"].values.astype("str"),
+    data["polarity"].values.astype("int"),
+    flags.BATCH_SIZE,
+    vectorize_layer,
+)
+mlm_train_data = IMDBReviewsDataGenerator(
+    data["text"].values.astype("str"),
+    None,
+    flags.BATCH_SIZE,
+    vectorize_layer,
+    model_type="mlm",
+)
+
+"""
+## Create BERT Model From Scratch
+"""
+
+
+class BERT:
+    def __init__(self, flags):
+        self.flags = flags
+
+    def __call__(self, inputs):
+        # embedding layer
+        embedding_layer = tf.keras.layers.Embedding(flags.VOCAB_SIZE, flags.EMBED_DIM)(
+            inputs
+        )
+
+        # query, value, key for attention
+        query = tf.keras.layers.Dense(flags.EMBED_DIM)(embedding_layer)
+        value = tf.keras.layers.Dense(flags.EMBED_DIM)(embedding_layer)
+        key = tf.keras.layers.Dense(flags.EMBED_DIM)(embedding_layer)
+
+        # multi headed self-attention
+        attention_output = tf.keras.layers.MultiHeadAttention(
+            num_heads=flags.NUM_HEAD, key_dim=flags.EMBED_DIM // flags.NUM_HEAD
+        )(query, value, key)
+        attention_output = tf.keras.layers.Dropout(0.1)(attention_output)
+        attention_output = tf.keras.layers.LayerNormalization(epsilon=1e-6)(
+            embedding_layer + attention_output
+        )
+
+        # feed-forward leyer
+        ffn = tf.keras.Sequential(
+            [
+                tf.keras.layers.Dense(flags.FF_DIM, activation="relu"),
+                tf.keras.layers.Dense(flags.EMBED_DIM),
+            ]
+        )
+        ffn_output = ffn(attention_output)
+        ffn_output = tf.keras.layers.Dropout(0.1)(ffn_output)
+        sequence_output = tf.keras.layers.LayerNormalization(epsilon=1e-6)(
+            attention_output + ffn_output
+        )
+
+        pooled_output = tf.keras.layers.GlobalMaxPooling1D()(ffn_output)
+
+        return sequence_output, pooled_output
+
+
+def build_mlm_model():
+    tf.keras.backend.clear_session()
+    inputs = tf.keras.layers.Input((flags.MAX_LEN,), dtype=tf.int32)
+    sequence_output, pooled_output = BERT(flags)(inputs)
+    outputs = tf.keras.layers.Dense(flags.VOCAB_SIZE)(sequence_output)
+    mlm_model = tf.keras.Model(inputs, outputs)
+
+    loss_fn = tf.keras.losses.SparseCategoricalCrossentropy(
+        reduction=tf.keras.losses.Reduction.NONE, from_logits=True
```

Comment:
Why no reduction? since you're using compile/fit the results will be averaged for reduction anyway (since some form of reduction is necessary).